geomstats/learning/riemannian_mean_shift.py

```diff
@@ -0,0 +1,198 @@
+    """Class for Riemannian Mean Shift algorithm on manifolds.
+
+    Mean Shift is a procedure for locating the maxima - the modes of a
+    density function given discrete data sampled from that function. It is
+    an iterative method for finding the centers of a collection of clusters.
+
+    Following implementation assumes a flat kernel method.
+
+    Parameters
+    ----------
+    manifold : object of class RiemannianManifold
+        Geomstats Riemannian manifold on which the
+        Riemannian Mean Shift algorithm is to be applied.
+    metric : object of class RiemannianMetric
+        Geomstats Riemannian metric associated to the space used.
+    bandwidth : float
+        Size of neighbourhood around each center. All points in 'bandwidth'
+        size around center are considered for calculating new mean centers.
+    tol : float
+        Stopping condition. Computation of subsequent mean centers is stopped
+        when the distance between them is less than 'tol'.
+        Optional, default : 1e-2.
+    n_centers : int
+        Number of centers.
+        Optional, default : 1.
+    n_jobs : int
+        Number of parallel threads to be initiated for parallel jobs.
+        Optional, default : 1.
+    max_iter : int
+        Upper bound on total number of iterations for the centers to converge.
+        Optional, default : 100.
+    init_centers : str
+        Initializing centers, either from the given input points or
+        random points uniformly distributed in the input manifold.
+        Optional, default : "from_points".
+    kernel : str
+        Weighing function to assign kernel weights to each center.
+        Optional, default : "flat".
+    """
+
+    def __init__(
+        self,
+        manifold,
+        metric,
+        bandwidth,
+        tol=1e-2,
+        n_centers=1,
+        n_jobs=1,
+        max_iter=100,
+        init_centers="from_points",
+        kernel="flat",
+        **frechet_mean_kwargs
+    ):
+        self.manifold = manifold
+        self.metric = metric
+        self.bandwidth = bandwidth
+        self.tol = tol
+        self.mean = FrechetMean(self.metric, **frechet_mean_kwargs)
+        self.n_centers = n_centers
+        self.n_jobs = n_jobs
+        self.max_iter = max_iter
+        self.init_centers = init_centers
+        self.kernel = kernel
+        self.centers = None
+
+    def dist_intersets(self, points_a, points_b, **joblib_kwargs):
+        """
+        Parallel computation of distances between two sets of points.
+
+        Parameters
+        ----------
+        points_a : array-like, shape=[..., n_features]
+            Clusters of points.
+        points_b : array-like, shape=[..., n_features]
+            Clusters of points.
+        """
+        n_a, n_b = points_a.shape[0], points_b.shape[0]
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_dist(x, y):
+            """
+            Riemannian metric between points x & y.
```

Comment:
This is actually `Riemannian distance` between points x, and y. The distance comes from the Riemannian metric, but they are two different concepts.

Also: the one-liner docstring should come after the """, like:
```
"""Compute Riemannian distances between x and y.

Parameters
------------
...
```